Add efi variable to augment /proc/cmdline

keszybz opened the pull request.

keszybz commented:
It seems that the first patch for cg_unified_flush()/cg_unified_update() is wrong.

keszybz commented:
I reworked the first patch. It passes on its own, so the rest should now pass too.

evverx commented:
Regarding `test-journal-flush`, it also failed in https://github.com/systemd/systemd/pull/13250, which most likely means it has nothing to do with this PR but as far as I can tell, the test isn't exactly a unit test in the sense that it relies on the journal not being corrupted and it's probably going to fail until the image used on Ubuntu CI gets updated.

poettering commented on src/basic/efivars.c:
```diff
@@ -222,4 +222,29 @@ int efi_set_variable_string(sd_id128_t vendor, const char *name, const char *v)
+                return 0;
+        }
+
+        r = efi_get_variable_string(EFI_VENDOR_LOADER, "SystemdOptions", line);
```
hmm, shouldn't this maybe something the loader appends, not userspace?

i mean, in particular as you put this in the loader UEFI UUID namespace?

The capitalization of "SystemdOptions" is a bit weird...

Shouldn't this be KernelCommandLineAppend or so, and then simply be appended by the boot loader or the EFI stub? I mean, otherwise people would try to add "audit=0 selinux=0" and then be very annoyed and surprised if it doesn't work.

Also, I am pretty sure this should also come in a oneshot flavour, so that you can append "fsck.mode=force" for the next boot only.

keszybz replied:
> hmm, shouldn't this maybe something the loader appends, not userspace?

I wanted this to work independently of the loader. In particular, I'm thinking about providing an alternative way to say 'systemd.uniified-cgroup-hierarchy=no' in F31 after we switch the default. For the loader, we already have a well-established way to set options. I guess it'd make sense to add something like this too sd-boot too, but for me that's an independent issue.

> as you put this in the loader UEFI UUID namespace?

Good point. I'll move it to a different namespace.

> The capitalization of "SystemdOptions" is a bit weird...

Suggestions?

>  KernelCommandLineAppend

As you say, this would then confuse people because it's only about systemd in this patch. I think KernelCommandLineAppend (or Prepend?) could be used for the sd-boot option.

> onshot flavour

OK, will add.

keszybz replied:
Actually, what does `EFI_VENDOR_LOADER` and `EFI_VENDOR_GLOBAL` mean? What namespace would be reasonable for "systemd options"?

poettering replied:
introduce a new namespace for that. It's just a uuid you can make up.

I am not sure doing a new efi var that feels like a kernel cmdline string but also totally isnt is the best way forward.

I see two options:

1. This could be an option read by the EFI loader stub, and not the EFI boot menu, and is added implicitly to the kernel cmdline.

2. Instead of adding a string formatted option, instead add explicitly settings, i.e. EFI_VENDOR_SYSTEMD:DefaultUnit= for systemd.unit= EFI_VENDOR_SYSTEMD:FileSystemCheck= for systemd.fsck= and so on. i.e. make them high-level variables, and only expose the ones that have the strong usecase. In this case they feel distinct enough not to be confused with a kernel cmdline option.

keszybz replied:
> ntroduce a new namespace for that. It's just a uuid you can make up.

That's what I did.

> This could be an option read by the EFI loader stub, and not the EFI boot menu, and is added implicitly to the kernel cmdline.

This doesn't really solve the goal: I want to have to have a way to amend *systemd* options independently of the boot loader used. And I want this to work "now", i.e. when we are pushing cgroups v2 in Fedora. Some future solution that requires discussion and changes in other projects doesn't really help.

> Instead of adding a string formatted option, instead add explicitly settings, i.e. EFI_VENDOR_SYSTEMD:DefaultUnit= for systemd.unit= EFI_VENDOR_SYSTEMD:FileSystemCheck= for systemd.fsck=

We already expose only specific settings for overriding on the kernel command line, so I think the list that is overridable through EFI should be the same. Otherwise things would be confusing and harder to maintain for no good reason.

OK, I'll try to amend the patch like this.

keszybz commented:
Updated:
- renamed efi_options_variable → efi_systemd_options_variable
- new EFI_VENDOR_SYSTEMD namespace for the variable

keszybz commented:
Rebased and updated (slightly):
- the docs are now very verbose that the EFI variable is not a replacement for the /proc/cmdline and only applies to systemd

@poettering wrote:
> Instead of adding a string formatted option, instead add explicitly settings, i.e. EFI_VENDOR_SYSTEMD:DefaultUnit= for systemd.unit= EFI_VENDOR_SYSTEMD:FileSystemCheck= for systemd.fsck= and so on. i.e. make them high-level variables, and only expose the ones that have the strong usecase. 

I tried that, but this much more annoying (to implement and to use). The biggest issue is that I think we should simply allow *all* variables that are parsed on the kernel command line to be parsed in this manner. If we start with some that "have a strong use case", then inevitably people will ask for additional ones. We'll be playing whack-a-mole, and there will no clear way to figure out if a given systemd version supports a given variable or not. I think that if we decided to implement command-line parsing for a given variable, than that's a strong-enough use case to also support overriding it through the EFI variable. But once we accept that we allow multiple variables, making a separate EFI variable for each and every one is ugly. In particular, there are variables which modify the same setting, e.g. `3` and `systemd.unit=` and `single`. If we allowed them to be set through separate variables, some way to resolve the order needs to be designed. If we ever stop looking at some variable, or there is a variable which is supported in a newer systemd version set by the user, there is no way to "clean up", and there is no way to tell the user that a variable is unsupported. Also, a separate brand new naming scheme is necessary to use for the variable names (systemd.fsck↔FileSystemCheck, …). All in all, I think that the current approach is much more user friendly without bringing unnecessary complexity. As I understand, the motivation to split into separate variables is to avoid confusion between this and the kernel command line. I think that the additions I added to the docs now are enough to serve this purpose.

keszybz commented:
This was reviewed a few times, and although there were questions about the user interface, and I think they have been answered. I'm happy with how this works. If nobody objects, I'll merge this.
